urlbd77b745af_KokaKiwi_MCLauncher_tgz-pJ8-net_contrapunctus_lzma_BenchmarkJ8/src/SevenZip/Compression/LZ/BinTree.java

```diff
@@ -0,0 +1,431 @@
+    static final int kHash3Size          = 1 << 16;
+    static final int kBT2HashSize        = 1 << 16;
+    static final int kStartMaxLen        = 1;
+    static final int kHash3Offset        = kHash2Size;
+    static final int kEmptyHashValue     = 0;
+    static final int kMaxValForNormalize = (1 << 30) - 1;
+    
+    int              kNumHashDirectBytes = 0;
+    int              kMinMatchCheck      = 4;
+    int              kFixHashSize        = kHash2Size + kHash3Size;
+    
+    public void SetType(int numHashBytes)
+    {
+        HASH_ARRAY = numHashBytes > 2;
+        if (HASH_ARRAY)
+        {
+            kNumHashDirectBytes = 0;
+            kMinMatchCheck = 4;
+            kFixHashSize = kHash2Size + kHash3Size;
+        }
+        else
+        {
+            kNumHashDirectBytes = 2;
+            kMinMatchCheck = 2 + 1;
+            kFixHashSize = 0;
+        }
+    }
+    
+    
+    public void Init() throws IOException
+    {
+        super.Init();
+        for (int i = 0; i < _hashSizeSum; i++)
+        {
+            _hash[i] = kEmptyHashValue;
+        }
+        _cyclicBufferPos = 0;
+        ReduceOffsets(-1);
+    }
+    
+    
+    public void MovePos() throws IOException
+    {
+        if (++_cyclicBufferPos >= _cyclicBufferSize)
+        {
+            _cyclicBufferPos = 0;
+        }
+        super.MovePos();
+        if (_pos == kMaxValForNormalize)
+        {
+            Normalize();
+        }
+    }
+    
+    public boolean Create(int historySize, int keepAddBufferBefore,
+            int matchMaxLen, int keepAddBufferAfter)
+    {
+        if (historySize > kMaxValForNormalize - 256)
+        {
+            return false;
+        }
+        _cutValue = 16 + (matchMaxLen >> 1);
+        
+        final int windowReservSize = (historySize + keepAddBufferBefore
+                + matchMaxLen + keepAddBufferAfter) / 2 + 256;
+        
+        super.Create(historySize + keepAddBufferBefore, matchMaxLen
+                + keepAddBufferAfter, windowReservSize);
+        
+        _matchMaxLen = matchMaxLen;
+        
+        final int cyclicBufferSize = historySize + 1;
+        if (_cyclicBufferSize != cyclicBufferSize)
+        {
+            _son = new int[(_cyclicBufferSize = cyclicBufferSize) * 2];
+        }
+        
+        int hs = kBT2HashSize;
+        
+        if (HASH_ARRAY)
+        {
+            hs = historySize - 1;
+            hs |= hs >> 1;
+            hs |= hs >> 2;
+            hs |= hs >> 4;
+            hs |= hs >> 8;
+            hs >>= 1;
+            hs |= 0xFFFF;
+            if (hs > 1 << 24)
+            {
+                hs >>= 1;
+            }
+            _hashMask = hs;
+            hs++;
+            hs += kFixHashSize;
+        }
+        if (hs != _hashSizeSum)
+        {
+            _hash = new int[_hashSizeSum = hs];
+        }
+        return true;
+    }
+    
+    public int GetMatches(int[] distances) throws IOException
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 26. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 144-274 into a separate method.